lib/src/index.rs

```diff
@@ -125,16 +159,30 @@ pub trait MutableIndex {
     fn merge_in(&mut self, other: &dyn ReadonlyIndex);
 }
 
+/// Defines the interface for types that provide an index of the commits in a
+/// repository by [`ChangeId`].
 pub trait ChangeIdIndex: Send + Sync {
-    /// Resolve an unambiguous change ID prefix to the commit IDs in the index.
+    /// Resolves a change ID prefix against all change IDs in the index, and
+    /// returns:
     ///
-    /// The order of the returned commit IDs is unspecified.
+    ///   - `NoMatch` if the prefix does not match any change IDs in the index,
+    ///     or if all commits for the change ID it matches are hidden.
+    ///   - `SingleMatch` if the prefix resolves unambiguously and points to one
+    ///     or more visible commits. The order of the returned commits is
+    ///     unspecified.
+    ///   - `AmbiguousMatch` if the prefix matches multiple change IDs. Since
+    ///     the search space is all change IDs in the index, `AmbiguousMatch`
+    ///     may be returned even if the prefix is unambiguous against the set of
+    ///     visible commits.
     fn resolve_prefix(&self, prefix: &HexPrefix) -> PrefixResolution<Vec<CommitId>>;
 
     /// This function returns the shortest length of a prefix of `key` that
-    /// disambiguates it from every other key in the index.
+    /// disambiguates it from every other key in the index, including hidden
+    /// keys. Since it disambiguates among all keys, the returned length is
+    /// usually a few digits longer than the minimum length necessary to
+    /// disambiguate among the visible keys only.
```

Comment:
I think this comment should still say what it said before, i.e. this:

```
    /// This function returns the shortest length of a prefix of `key` that
    /// disambiguates it from every other key in the index.
```

Then put the comment about including hidden commits on `ChangeIdIndexImpl` as a `//` style comment.

Does that make sense?

Reply:
As a `//` comment on the trait or on the implementation? It doesn't seem implementation specific, but I guess it is supplementary info we don't need to include in the doc comment.

Reply:
It is specific to the default implementation. I don't think we want to specify that all backends should consider hidden commits. Our backend at Google actually always returns the full length because we don't have an efficient way of finding the shortest prefix :)

Reply:
Oh I see. Ok I've moved the comment. So I guess the important thing is that this disambiguates against every key in the index, which might not be every key in the repo? For example, the index could only have entries for visible keys (commits)?

Sorry for all the questions, I just want to make sure I understand and that we've resolved any ambiguity.

Reply:
I've restored the documentation in composite.rs so that it's basically the same as it is at HEAD.

As I said in the other thread, I think there is a little too much ambiguity about which commits are considered by `ChangeIdIndex::resolve_prefix` and `ChangeIdIndex::shortest_unique_prefix_len`, but maybe it is just my lack of experience with the project.